### PR TITLE
Add concatParse for chained parsing

### PR DESCRIPTION
**Pull request by harendra-kumar**
Iterate parsers on a stream such that the next parser depends on the previous parser, generating an output stream from the intermediate parse results.

### CODE REVIEW

**Comment by harendra-kumar** [comment]
GHC 7.10.3 build fails with out of memory:

```
ghc: out of memory (requested 2097152 bytes) 
...
cabal: Failed to build lib:lib-prelude from streamly-benchmarks-0.0.0 (which
is required by bench:serial-o-n-stack from streamly-benchmarks-0.0.0,
bench:serial-o-n-space from streamly-benchmarks-0.0.0 and others).
```

File: src/Streamly/Internal/Prelude.hs [hunk 1]
```diff
@@ -3138,25 +3142,72 @@ concatMapTreeYieldLeavesWith combine f = concatMapLoopWith combine f yield
...
+    -> Parser m a x
+    -> t m a
+    -> t m b
+parseManyTill = undefined
+
+-- | Iterate a parser generating function on a stream. The initial value @b@ is
+-- used to generate the first parser, the parser is applied on the stream and
+-- the result is used generate the next parser and so on.
+--
+-- >>> S.toList $ S.parseIterate (\b -> PR.take 2 (FL.mconcatTo b)) 0 $ S.fromList [1..10]
```
Comment (pranaysashank) [inline src/Streamly/Internal/Prelude.hs:3194]: Integers do not have a `Monoid` instance AFAIK

File: src/Streamly/Internal/Data/Fold.hs [hunk 2]
```diff
@@ -698,13 +733,20 @@ index = genericIndex
...
 -- @since 0.7.0
 {-# INLINABLE head #-}
 head :: Monad m => Fold m a (Maybe a)
 head = _Fold1 const
 
 -- | Returns the first element that satisfies the given predicate.
 --
+-- @
+-- find p = fmap getFirst $
+--     foldMap (\x -> First (if p x then Just x else Nothing))
```
[the remark this answers is not in the document]
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Fold.hs]: fixed

File: src/Streamly/Internal/Prelude.hs (hunk 1 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Prelude.hs]: fixed the example.

**Pull request merged**